RDoc-1355 Counters documentation (#900)
ravendb/docs: Documentation/4.1/Raven.Documentation.Pages/client-api/session/counters/counters-and-other-features.markdown

  * [Counters and Smuggler](../../../client-api/session/counters/counters-and-other-features#counters-and-smuggler)
  * [Counters and Changes API](../../../client-api/session/counters/counters-and-other-features#counters-and-changes-api)
  * [Counters and Ongoing Tasks (Backup, External replication, ETL, Subscriptions)](../../../client-api/session/counters/counters-and-other-features#counters-and-ongoing-tasks)
  * [Counters and Other Features: summary](../../../client-api/session/counters/counters-and-other-features#counters-and-other-features-summary)
  * [Including Counters](../../../client-api/session/counters/counters-and-other-features#including-counters)
{NOTE/}

---

{PANEL: }

### Counters and Indexing  

Indexing Counters can speed-up finding them and the documents that contain them.  
RavenDB supports Counters-indexing **by name**, but **not by value**.  

* **Counter-Names indexing**  
    Re-indexing due to Counter-name modification is normally rare enough to pause no performance issues.  
    To index a document's Counters by name, use [CounterNamesFor](../../../indexes/indexing-counters#indexes--indexing-counters).  
  
* **Counter-Values indexing**  

>     Indexing by Counter-value is **not** provided, because it may initiate re-indexing intensive enough to hurt performance.  

Author: done

* **Counter-Values indexing**  
    Indexing by Counter-value is **not** provided, changing the Counter's value will not trigger indexing of the document.  
    Counters are designed for high-throughput scenarios and cannot afford the re-indexing cost of each change.

---

###Counters and Queries  

You can create queries using code, or send the server raw queries for execution.  
Either way, you can **query documents by Counter-names** but **not** filter results by Counter values.  
This is because queries are generally [based on indexes](../../../start/getting-started#example-iii---querying), and Counter values are [not indexed](../../../client-api/session/counters/counters-and-other-features#counters-and-indexing).  

> 

Author: is this clear enough?

Create queries **using code**, or send the server **raw queries** for execution.  

* Either way, you can query Counters **by name** but **not by value**.  
  This is because queries are generally [based on indexes](../../../start/getting-started#example-iii---querying), and Counter values are [not indexed](../../../client-api/session/counters/counters-and-other-features#counters-and-indexing).  
* Counter values **can** be [projected](../../../../indexes/querying/projections) from query results, as demonstrated in the following examples.  
  This way a client can get Counter values from a query without downloading whole documents.

> 

Reviewer: Okay

* Use [Session.Query](../../../client-api/session/querying/how-to-query#session.query) to code queries yourself.  
   * **Returned Counter Value**: **Accumulated**  
     A Counter's value is returned as a single sum, with no specification of the Counter's value on each node.
     {CODE counters_region_query@ClientApi\Session\Counters\Counters.cs /}

* Use [RawQuery](../../../client-api/session/querying/how-to-query#session.advanced.rawquery) to send the server raw RQL expressions for execution.  
   * Use the `counter` method.  
     **Returned Counter Value**: **Accumulated**

  * Use the `counterRaw` method.  
     **Returned Counter Value**: **Distributed**  

>      A Counter's value is returned as a series of values, each maintained by one of the nodes.

Reviewer: ```suggestion
     A Counter's value is returned as a series of values, each maintained by one of the nodes. This is very rarely needed and it isn't expected that you'll use that in your applications.
```

A [document revision](../../../client-api/session/revisions/what-are-revisions) stores all the document's Counters' names and values.  
 
* **Stored Counter Values**: **Accumulated**  
  A revision stores a Counter's value as a single sum, with no specification of the Counter's value on each node.  

* Revisions-creation can be initiated by **Counter-name modification**.  
   * When the Revisions feature is enabled, the creation or deletion of a Counter initiates the creation of a new document revision.  
   * Counter **value** modifications do **not** cause the creation of new revisions.  

---

###Counters and Smuggler  

[Smuggler](../../../client-api/smuggler/what-is-smuggler) is a DocumentStore property, that can be used 
to [export](../../../client-api/smuggler/what-is-smuggler#databasesmugglerexportoptions) chosen 
database items to an external file or to [import](../../../client-api/smuggler/what-is-smuggler#databasesmugglerimportoptions) 
database items from an existing file into the database.  

* **Transferred Counter Value**: **Distributed**  
  Smuggler transfers the entire series of values that the different nodes maintain for a Counter.  

> * To make Smuggler handle Counters, set the `OperateOnTypes` parameter to `DatabaseItemType.Counters`.  

Reviewer: ```suggestion
* To make Smuggler handle Counters, set the `OperateOnTypes` parameter to include `DatabaseItemType.Counters`.  
```

Reviewer: Note that this is a multi value flag

Author: * To make Smuggler handle Counters, include `DatabaseItemType.Counters` in `OperateOnTypes`'s list of entities to import or export.  
  {CODE smuggler_options@ClientApi\Session\Counters\Counters.cs /}

* **Counters** and the **External Replication task**  
    The ongoing [external replication](../../../studio/database/tasks/ongoing-tasks/external-replication-task) task replicates all data, including Counters.  
    * **Replicated Counter Value**: **Distributed**  
        Counters are replicated along with their values on all nodes.  
    * Replication can be initiated by both **Counter-name update** _and_ **Counter-value modification**.  

* **Counters** and the **ETL task**  
    [ETL](../../../server/ongoing-tasks/etl/basics) is used to export data from RavenDB to an external (either Raven or SQL) database.  
    * [SQL ETL](../../../server/ongoing-tasks/etl/sql) is **not supported**.  
      Counters cannot be exported to an SQL database over SQL ETL.  
    * [RavenDB ETL](../../../server/ongoing-tasks/etl/raven) **is supported**.  
      Counters [are](../../../server/ongoing-tasks/etl/raven#counters) exported over RavenDB ETL.  
      * Export can be initiated by both **Counter-name update** _and_ **Counter-value modification**.  
      * **Exported Counter Value**: **Distributed**  
        Counters are exported along with their values on all nodes.  

>       * **Default behavior**: When an ETL script is not provided, Counters are exported by default.  

Reviewer: Need to mention how you can ETL counters with script

Author: done

| [Backup Task](../../../client-api/session/counters/counters-and-other-features#counters-and-ongoing-tasks) | _Time Schedule_ | _Distributed_ |
| [RavenDB ETL Task](../../../client-api/session/counters/counters-and-other-features#counters-and-ongoing-tasks) | _Document Change_, <br> _Countrer Value Modification_ | _Distributed_ |
| [External Replication task](../../../client-api/session/counters/counters-and-other-features#counters-and-ongoing-tasks) | _Document Change_, <br> _Countrer Value Modification_ | _Distributed_ |
| [Data Subscriptions Update Task](../../../client-api/session/counters/counters-and-other-features#counters-and-ongoing-tasks) | _Document Change_ | |
| [Changes API](../../../client-api/session/counters/counters-and-other-features#counters-and-changes-api) | _Countrer Value Modification_ | _Accumulated_ |
| [Revision creation](../../../client-api/session/counters/counters-and-other-features#counters-and-revisions) | _Document Change_ | _Accumulated_ |

{NOTE/}

---

###Including Counters  
You can [include](../../../client-api/how-to/handle-document-relationships#includes) Counters while loading a document.  

> An included Counter is cached, so it can be immediately retrieved when needed.  

Reviewer: ```suggestion
An included Counter is retrieved in the same request as its owner document and is held by the sesion, so it can be immediately retrieved when needed with no additional remote calls.
```

Author: fixed, thanks
